File: backend/schemas/team.py
"""
Pydantic Schemas for Team Entity

These schemas handle validation and serialization for Team-related API operations.
Team data is sourced from Excel sheets (Equipes, Preferences_Gymnases).
"""

from pydantic import BaseModel, ConfigDict
from typing import Optional, List
from datetime import datetime


class TeamBase(BaseModel):
    """
    Base schema with common fields for Team.
    
    Data sources:
    - nom, poule: From Excel sheet "Equipes"
    - institution, numero_equipe, genre: Extracted from team name
    - horaires_preferes: From Excel column "Horaire_Prefere"
    - lieux_preferes: From Excel sheet "Preferences_Gymnases" (by institution)
    """
    # Informations équipe (depuis feuille Excel "Equipes")
    nom: str
    institution: Optional[str] = None
    numero_equipe: Optional[str] = None
    genre: Optional[str] = None
    poule: Optional[str] = None
    
    # Préférences (depuis Excel ou calculées)
    horaires_preferes: Optional[List[Optional[str]]] = None  # Depuis "Horaire_Prefere"
    lieux_preferes: Optional[List[Optional[str]]] = None     # Depuis feuille "Preferences_Gymnases"


class TeamCreate(TeamBase):
    """
    Schema for creating a new team.
    
    Adds project_id to link team to a project.
    """
    project_id: int


class TeamUpdate(BaseModel):
    """
    Schema for partial team updates (PATCH).
    
    All fields are optional to support partial updates.
    """
    # Tous champs optionnels
    nom: Optional[str] = None
    institution: Optional[str] = None
    numero_equipe: Optional[str] = None
    genre: Optional[str] = None
    poule: Optional[str] = None
    horaires_preferes: Optional[List[Optional[str]]] = None
    lieux_preferes: Optional[List[Optional[str]]] = None


class TeamResponse(BaseModel):
    """
    Schema for team API responses.
    
    Includes all fields plus id, project_id, and timestamp.
    Uses from_attributes=True for automatic ORM → Pydantic conversion.
    
    Note: All fields redefined to ensure proper ORM conversion.
    """
    id: int
    project_id: int
    
    # Informations équipe
    nom: str
    institution: Optional[str] = None
    numero_equipe: Optional[str] = None
    genre: Optional[str] = None
    poule: Optional[str] = None
    
    # Préférences
    horaires_preferes: Optional[List[Optional[str]]] = None
    lieux_preferes: Optional[List[Optional[str]]] = None
    
    # Timestamp
    created_at: datetime
    
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/team.ts
/**
 * Types pour les équipes.
 * 
 * Source Excel : Feuille "Equipes" avec colonnes :
 * - Nom, Institution, Numéro équipe, Genre, Poule
 * 
 * Correspond au schéma Pydantic TeamResponse.
 */

export interface Team {
  id: number
  project_id: number
  
  // Identité équipe (depuis feuille Equipes)
  nom: string
  institution: string | null
  numero_equipe: string | null  // String car peut contenir "1", "2", "A", "B", etc.
  genre: string | null          // "Garçons", "Filles", "Mixte"
  poule: string | null          // "P1", "P2", "P3", etc.
  
  // Préférences (depuis feuille Equipes ou Preferences_Gymnases)
  horaires_preferes: string[] | null  // Ex: ["Mercredi 14:00", "Vendredi 18:00"]
  lieux_preferes: string[] | null     // Ex: ["Gymnase A", "Gymnase B"]
  
  // Timestamps
  created_at: string  // ISO 8601
}

/**
 * Données pour créer une nouvelle équipe.
 * 
 * Correspond au schéma Pydantic TeamCreate.
 */
export interface TeamCreate {
  project_id: number
  nom: string
  institution?: string | null
  numero_equipe?: string | null
  genre?: string | null
  poule?: string | null
  horaires_preferes?: string[] | null
  lieux_preferes?: string[] | null
}

/**
 * Données pour mettre à jour une équipe (PATCH).
 * 
 * Correspond au schéma Pydantic TeamUpdate.
 */
export interface TeamUpdate {
  nom?: string
  institution?: string | null
  numero_equipe?: string | null
  genre?: string | null
  poule?: string | null
  horaires_preferes?: string[] | null
  lieux_preferes?: string[] | null
}
